text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,409 @@
+   dependencies in offline environments.
+1. Via the platform. Platforms may choose to offer enhanced functionality to
+   more easily distribute dependencies. In the end, the platform just needs to
+   ensure that the dependency metadata is available to the buildpack in the
+   required location.
+
+In addition to the flexibility of where the dependencies originate, this
+proposal also provides flexibility in how those dependencies are managed such
+as floating them so the latest versions are always available, pinning to a
+specific set of dependencies or even pinning and including the dependencies
+with the metadata.
+
+## Dependency Assets
+This proposal does not impact how dependencies are being distributed and it
+supports the current methods referring to remote locations, like upstream
+projects, or using the Deps Server. The dependency metadata just needs to point
+to the location from which the dependency should be fetched as is currently
+being done with the buildpack.toml dependency metadata.
+
+For builders, buildpacks, or platforms that would like to inject dependency
+assets directly into the build container, perhaps to support offline builds, we
+propose defining `BP_DEPENDENCY_ASSET_ROOT` which defaults to
+`/platform/deps/assets` as the location where the actual dependency assets
+should be located (again, the intent is to pilot and try out this or possibly
+other locations, eventually proposing an RFC with Cloud-Native buildpacks to
+standardize the location).
+
+In this way, one could have dependency metadata that uses `file://` URLs to
+refer to a stable location, i.e. `/platform/deps/assets/…`. These dependencies
+could then be accessed by the buildpack directly, and the buildpack doesn’t
+need to care how they were added to the container.
+
+This document does not specify how the dependency asset folder should be
+provided to buildpacks but here are some possibilities:
+
+1. External. The dependency assets can be managed outside of the buildpack
+   lifecycle. This allows for users to manually pull in the dependency assets
+   they would like when they would like them. It also outlives the buildpack
+   lifecycle so assets could be shared across builds for greater caching
+   benefit. The assets can then be mapped into the build container at the
+   specified location using a volume mount to `pack build`.
+1. Via builder. The builder could come with assets included under
+   `/platform/deps/assets`. This would likely require a builder with many
+   buildpacks to update very frequently though, so may not be the best idea for
+   the average case, however, it could be very useful as a way to distribute
+   dependencies in offline environments.
+1. Via the platform. Platforms may choose to offer enhanced functionality to
+   more easily distribute dependencies. In the end, the platform just needs to
+   ensure that the dependency metadata is available to the buildpack in the
+   required location.
+
+## Implementation
```

Comment:
We should actually start with a couple of explorations of different use cases and clarify those details they I suggest we leave out of here.
Maybe we should have a list of explorations right here and invite all Paketo maintainers to have a look at there dependency handling and add explorations for their use cases.

```suggestion
## Implementation
### Exploration
We propose to start with a larger exploration phase in order to detail out the contracts between dependency installing buildpacks and dependency providers.
E.g.
- Could we have a single JVM buildpack and defer vendor variants to externally provided dependency metadata for e.g. Bellsoft, SapMachine, Microsoft, etc.?
- How can we keep multiple version lines of dependencies (Java 8, 11, 17, 20 and similar for Node)
- What dependency versioning schemes exist and is there a common pattern besides semver (e.g. regex) that fits all other cases?
- How do we handle precedence if a builder comes with batteries included, the platform adds dependencies and the user tries to override some of the provided ones?
- ...
```